### PR TITLE
feat: allow clang-format to format all metal files in repo

### PR DESCRIPTION
**Pull request by NripeshN**
## Proposed changes

This change allows metal files to be formatted

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: mlx/backend/metal/kernels/ternary.metal
Comment (awni) [inline mlx/backend/metal/kernels/ternary.metal:187]: This looks a little strange.. maybe worth adding a no format here.

File: mlx/backend/metal/kernels/unary.metal
Comment (awni) [inline mlx/backend/metal/kernels/unary.metal:59]: And here.

File: mlx/backend/metal/kernels/unary.metal
Comment (awni) [inline mlx/backend/metal/kernels/unary.metal:136]: and here.

File: mlx/backend/metal/kernels/steel/gemm/kernels/steel_gemm.metal
Comment (awni) [inline mlx/backend/metal/kernels/steel/gemm/kernels/steel_gemm.metal:195]: this looks bad to me, guard with no formatting

File: mlx/backend/metal/kernels/steel/gemm/kernels/steel_gemm_addmm.metal
Comment (awni) [inline mlx/backend/metal/kernels/steel/gemm/kernels/steel_gemm_addmm.metal:407]: Same here.

File: mlx/backend/metal/kernels/steel/gemm/kernels/steel_gemm_addmm.metal
Comment (awni) [inline mlx/backend/metal/kernels/steel/gemm/kernels/steel_gemm_addmm.metal:439]: And here.

**Review by awni — CHANGES_REQUESTED** [review]
I'd actually be in favor of adding formatting to metal code, so thanks for starting that. There are several places where we instantiate a bunch of kernels with the preprocessor and the formatted makes it very hard to read. Would you mind guarding those sections by turning the formatter off?

File: .pre-commit-config.yaml
Comment (awni) [inline .pre-commit-config.yaml:2]: Why remove the repo/version here?

**Review by awni — COMMENTED** [review]
Left a question, let me know. Otherwise I think it looks good, and we can merge shortly.

File: .pre-commit-config.yaml
Comment (NripeshN) [reply .pre-commit-config.yaml:2]: Clang-format's [pre-commit hook](https://github.com/pre-commit/mirrors-clang-format/blob/main/.pre-commit-hooks.yaml) does not have support to add metal files. Hence using custom hook seems like the solution

Comment (awni) [reply .pre-commit-config.yaml:2]: Could we keep the repo/version and instead specify the extensions in a `.clang-format` config?

I'm worried this is a bit brittle / relies on people having the same version of clang installed.

Comment (NripeshN) [reply .pre-commit-config.yaml]: I have created a custom pre-commit that will format all metal files now. Should solve the above problem

Comment (awni) [reply .pre-commit-config.yaml]: I don’t think that’s the right solution. In that case we have to rely on your hook to be updated and I’m not so comfortable having it as a dependency

Comment (NripeshN) [reply .pre-commit-config.yaml]: Maybe you can fork my repo in your organization and use the hook from there.

Comment (NripeshN) [reply .pre-commit-config.yaml]: The `.clang-format` configuration file is used to specify coding style preferences for `clang-format` but does not directly control which file extensions are processed by `clang-format`. The handling of specific file extensions and the invocation of `clang-format` on those files is typically managed outside of the `.clang-format` file, such as through script arguments or, in our case, pre-commit hook configurations.

Since the `mirrors-clang-format` pre-commit hook configuration we're using doesn't natively support `.metal` files through a simple configuration change in the `.pre-commit-config.yaml` or `.clang-format`, and you wish to avoid a `local` repo setup due to concerns about version consistency, we're in a bit of a bind. The pre-commit framework itself relies on matching files to hooks based on the configuration parameters provided in `.pre-commit-config.yaml`, not `.clang-format`.

Comment (NripeshN) [reply .pre-commit-config.yaml]: I have created a [PR](https://github.com/pre-commit/mirrors-clang-format/pull/31) in their repo.

Comment (NripeshN) [reply .pre-commit-config.yaml]: @awni they have closed the PR as the file I made changes to is being generated.
Given the context that the `.pre-commit-hooks.yaml` file is generated by a [GitHub Actions](https://github.com/pre-commit/mirrors-clang-format/blob/main/.github/workflows/main.yml) workflow, and the maintainers have indicated that my PR was closed because modifications should be made to the generation process rather than directly to the file, we'll need to modify the [GitHub Action](https://github.com/pre-commit/mirrors-clang-format/blob/main/.github/workflows/main.yml) itself to ensure `.metal` files are considered in the generated configuration.

The[ GitHub Action](https://github.com/pre-commit/mirrors-clang-format/blob/main/.github/workflows/main.yml) uses `pre-commit-mirror-maker` to automatically generate the pre-commit hook configuration for `clang-format`. The key part of the action where types are specified for the `clang-format` hook is in the `pre-commit-mirror` command. Since `.metal` files are not recognized as a type by pre-commit, we cannot directly add a new type for `.metal` files in the `--types-or` options.

However, considering the limitations, the goal to support Metal file formatting within the constraints of the existing system, we're somewhat at an impasse because the `pre-commit-mirror-maker` utility and the pre-commit framework itself rely on recognized file types.

Comment (awni) [reply .pre-commit-config.yaml]: >  Since .metal files are not recognized as a type by pre-commit

Can that be updated? It seems like they recognize `cuda`, `metal` is pretty standard.

Comment (NripeshN) [reply .pre-commit-config.yaml]: I had created [this](https://github.com/pre-commit/mirrors-clang-format/pull/32) PR in their repo and they are hesitant to merge it right away. Also I tried running the pre-commit hook using the commits SHA and it did not work. I think pre-commit itself does not support metal. Not entirely sure either.

Comment (awni) [reply .pre-commit-config.yaml]: Nice that looks like the right solution to me. Maybe we wait until they merge that and then we can come back to this.

Comment (awni) [reply .pre-commit-config.yaml]: @NripeshN should I leave this PR open until Metal is in the precommit? Are you planning to follow up there?

**Pull request closed without merging**
